Add desktop notifications

antonlindstrom opened the pull request:
This adds desktop notifications for Gone to be able to see when a
session has ended. The change introduces a message that will say for
instance that "Your working session has ended, time for a short break".
A Notifier interface has been added to make sure that the feature can be
extended with custom notifiers.

Desktop notifications is nice to have to be able to get an interruption
when a session is done.

Thanks!

antonlindstrom committed:
Add desktop notifications

This adds desktop notifications for Gone to be able to see when a
session has ended. The change introduces a message that will say for
instance that "Your working session has ended, time for a short break".
A Notifier interface has been added to make sure that the feature can be
extended with custom notifiers.

Desktop notifications is nice to have to be able to get an interruption
when a session is done.

cwarden commented on state/state.go:
```diff
@@ -99,7 +99,22 @@ func (s *State) Message() string {
+		return fmt.Sprintf("A %s is in progress", msgs[s.pattern[s.currentIdx]])
+	}
+
+	return fmt.Sprintf("Your %s has ended, time for a %s.", msgs[s.pattern[s.currentIdx]], msgs[s.pattern[s.currentIdx+1]])
```
We need to get `msgs[s.pattern[0]]` for the next message if `s.currentIdx == len(s.pattern)-1`.

antonlindstrom replied:
Thank you for the feedback and sorry for the delay in fixing it. The latest commit should have the changes requested as well as validation that it hopefully works. :bow:

antonlindstrom committed:
Add testing and support for circular state

This adds tests for the state status message and fixes some bugs with
circular state and if the state enters an invalid index.

antonlindstrom commented:
@guillaumebreton Any updates on this? Would you consider this change?

Thanks!

guillaumebreton reviewed (approved):
Thanks for the PR 😊 and sorry for the delayed review

antonlindstrom commented:
Awesome, thank you so much for this! :bowing_man: 

Have an amazing day!
